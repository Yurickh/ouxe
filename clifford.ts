import { EOL } from 'os'
import execa from 'execa'
import { TextDecoder } from 'util'
import { Readable } from 'stream'
import { EventEmitter } from 'events'
import AnsiParser from 'node-ansiparser'
import { AnsiTerminal } from 'node-ansiterminal'

interface CliffordOptions {
  readTimeout: number | false
  debug: boolean
  useBabelNode: boolean
  replacers: ((chunk: string) => string)[]
}

const defaultConfig = (command: string) => ({
  debug: false,
  readTimeout: 1000,
  useBabelNode: !command.endsWith('.js'),
  replacers: [],
})

const execaOptions = () => ({
  all: true,
  preferLocal: true,
})

function isRegExp(value: unknown): value is RegExp {
  return Object.prototype.toString.call(value) === '[object RegExp]'
}

function test(matcher: string | RegExp): (subject: string) => boolean
function test(matcher: string | RegExp, subject: string): boolean
function test(matcher: string | RegExp, subject?: string) {
  const applyMatcher = (subject: string) => {
    if (isRegExp(matcher)) {
      return matcher.test(subject)
    } else {
      return subject.includes(matcher)
    }
  }

  if (subject === undefined) {
    return applyMatcher
  }

  return applyMatcher(subject)
}

function commonLeadingString(first: string, second: string) {
  return first.slice(
    0,
    first.split('').findIndex((value, index) => value !== second[index]),
  )
}

const spawnNode = (command: string, args: string[]) =>
  execa('node', ['--', command, ...args], execaOptions())

const spawnBabelNode = (command: string, args: string[]) =>
  execa(
    'babel-node',
    ['--extensions', '.ts,.js', '--', command, ...args],
    execaOptions(),
  )

/**
 * Normalizes the chunk so its read the same accross platforms.
 * @param chunk A chunk of the readable stream of cli.all
 */
const normalizeChunk = (
  chunk: Uint8Array,
  replacers: ((chunk: string) => string)[],
) => {
  const newBuffer = new Uint8Array(
    Array.from(chunk).reduce((chunk, char) => {
      return [...chunk, ...(char === 10 ? [10, 13] : [char])]
    }, []),
  )

  return replacers.reduce(
    (string, replacer) => replacer(string),
    new TextDecoder().decode(newBuffer),
  )
}

interface ReaderConfig {
  replacers?: ((chunk: string) => string)[]
  debug?: boolean
}

class EventQueue {
  private readPointer: number
  private writePointer: number

  constructor(private emitter: EventEmitter, private event: string) {
    this.readPointer = 0
    this.writePointer = 0

    this.emitter.on(this.event, this.write)
  }

  private write() {
    ++this.writePointer
  }

  private isLagging() {
    return this.readPointer < this.writePointer
  }

  public dispose() {
    this.emitter.off(this.event, this.write)
  }

  public async next() {
    return new Promise<true>((resolve) => {
      if (this.isLagging()) {
        ++this.readPointer
        resolve(true)
      }

      this.emitter.once(this.event, () => {
        ++this.readPointer
        resolve(true)
      })
    })
  }
}

class Terminal {
  parser: AnsiParser
  terminal: AnsiTerminal

  constructor() {
    this.terminal = new AnsiTerminal(1000, 1000, Infinity)
    this.parser = new AnsiParser(this.terminal)
  }

  public read(): string {
    return this.terminal.toString()
  }

  public write(chunk: string): void {
    return this.parser.parse(chunk)
  }
}

class LineFeedEmitter extends EventEmitter {}

class Reader {
  screen: string
  terminal: Terminal
  debug: boolean
  lineFeedEmitter: LineFeedEmitter

  constructor(private stream: Readable, config: ReaderConfig) {
    this.screen = ''
    this.terminal = new Terminal()
    this.debug = config.debug ?? false
    // Turns out o que a gente quer aqui não é um emitter, e sim uma stream mesmo,
    // porque o emitter só vai notificar o pessoal que já está subscrito, mas a gente
    // quer _esperar_ o próximo evento, mesmo se ele já aconteceu
    this.lineFeedEmitter = new LineFeedEmitter()

    this.startListening(config.replacers)
  }

  private startListening(replacers: ((chunk: string) => string)[]) {
    this.stream.on('data', (chunk) => {
      normalizeChunk(chunk, replacers)
        .split(EOL)
        .forEach((chunkLine, index, array) => {
          // We want to setImmediate so each line feed is reacted to independently
          setImmediate(() => {
            this.terminal.write(
              chunkLine + (index + 1 === array.length ? '' : EOL),
            )
            this.lineFeedEmitter.emit('line')

            if (this.debug) {
              console.info('[current screen]\n', this.readScreen())
            }
          })
        })
    })
  }

  private updateScreen() {
    const currentScreen = this.screen
    const terminalScreen = this.readScreen()

    // We want to skip overriding the whole screen if the client cleans up the screen to
    // prepare the next print. Instead, we rely that the startsWith check will take care
    // of cleaning up if everything is changed.
    if (terminalScreen !== '') {
      this.screen = terminalScreen
    }

    if (terminalScreen.startsWith(currentScreen)) {
      return terminalScreen.slice(currentScreen.length)
    } else {
      return terminalScreen.slice(
        commonLeadingString(currentScreen, terminalScreen).length,
      )
    }
  }

  // TODO: this doesn't seem to work :T
  public untilClose() {
    return new Promise((resolve) => {
      this.stream.once('close', () => {
        resolve()
      })
    })
  }

  public readScreen() {
    // TODO: add types to terminal
    return this.terminal.read().trimRight()
  }

  /**
   * Method to wait until a line is printed with a given string.
   * @param matcher The string you're looking for, or a regex expression to match.
   */
  public async until(matcher: string | RegExp) {
    const eventQueue = new EventQueue(this.lineFeedEmitter, 'line')
    do {
      // We update the screen on every iteration so we get only the final chunk by the end
      const diff = this.updateScreen()
      if (test(matcher, diff)) {
        eventQueue.dispose()
        return diff
      }
    } while (await eventQueue.next())
  }

  /**
   * Method to find a line in the screen that matches the matcher.
   * Will search from bottom to top and return the first line that matches.
   * @param matcher The screen you're looking for, or a regex expression to match.
   */
  public async findByText(matcher: string | RegExp) {
    const screen = this.readScreen()

    if (!test(matcher, screen)) {
      return this.until(matcher)
    }

    return screen.split(EOL).reverse().find(test(matcher))
  }
}

export default function clifford(
  command: string,
  args: string[] = [],
  options: Partial<CliffordOptions> = {},
) {
  const optionsWithDefault: CliffordOptions = {
    ...defaultConfig(command),
    ...options,
  }

  const spawner = optionsWithDefault.useBabelNode ? spawnBabelNode : spawnNode
  const cli = spawner(command, args)

  const stringification = `[Clifford instance: running process for \`${command}\` with args \`${JSON.stringify(
    args,
  )}\` ]`

  const reader = new Reader(cli.all, {
    debug: optionsWithDefault.debug,
    replacers: optionsWithDefault.replacers,
  })

  return {
    // Although we don't need await here, it seems `write` might be async on windows
    type: async (string: string) => cli.stdin.write(`${string}\n`),
    read: () => cli.then(({ all }) => all),
    findByText: (matcher: string | RegExp) => reader.findByText(matcher),
    readScreen: () => reader.readScreen(),
    readUntil: (matcher: string | RegExp) => reader.until(matcher),
    untilClose: () =>
      Promise.race([
        reader.untilClose(),
        new Promise((resolve) => {
          cli.once('close', resolve)
        }),
      ]),
    kill: () => cli.cancel(),
    toString: () => stringification,
    toJSON: () => stringification,
  }
}
/**
 * Hey there Yurick from the future! I hope you're having a good time :)~
 * The last thing you noticed you needed to do here was to make so the read methods do it by
 * a per-line basis. That means you'll have to add some logic on the until method so it returns
 * the line that matched, and we skip adding the lines that came after that to the internal screen
 * representation.
 * This is the best solution because sometimes inquirer will yield two chunks or one chunk for the
 *  same strings, which makes it super hard to do snapshot testing.
 * Good luck! :hug:
 */
