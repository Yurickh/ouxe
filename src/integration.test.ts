import * as fs from 'fs-extra'
import * as path from 'path'
import clifford from '../clifford'
import reserveFile from './helpers/reserve-file'

// TODO: consider moving this to clifford
const clearColorMarkers = (string: string | undefined): string =>
  // eslint-disable-next-line no-control-regex
  string?.replace(/\x1b\[\d{0,3}[\w]/g, '').replace(/^\s/g, '')

const rootPath = (pathName: string): string =>
  path.join(__dirname, '..', pathName)

const runOuxe = (args: string[], options: object = {}) =>
  clifford('./src/index.ts', args, options)

describe('ouxe', () => {
  let returnFile: () => void

  beforeEach(() => {
    returnFile = () => undefined
  })

  afterEach(() => {
    returnFile()
  })

  it('runs help', async () => {
    const cli = runOuxe(['--help'])

    expect(await cli.read()).toMatchSnapshot()
  })

  it('prompts for a packager if none is identified', async () => {
    returnFile = reserveFile(rootPath('yarn.lock'))

    const cli = runOuxe(['prettier', '--skip-install'])

    const promptPackager = await cli.readLine()
    expect(promptPackager).toMatch(
      'Which package manager do you intend to use?',
    )
    // press enter for yarn
    await cli.type('')

    const promptWrite = await cli.readUntil(
      /Do you want to immediately run prettier.*/,
    )
    expect(promptWrite).toMatch(
      'Do you want to immediately run prettier on all files in the project?',
    )
  })

  describe('running prettier', () => {
    it('writes all files', async () => {
      const writableFile = path.join(__dirname, 'fixtures/write-prettier.js')
      returnFile = () => fs.removeSync(writableFile)

      fs.writeFileSync(
        writableFile,
        `
      export default function potato  () {
        let a,
        b,
        c;

        [a, b, c].forEach(x =>
         0)
      } 
      `,
      )

      const cli = runOuxe(['prettier', '--skip-install'])

      const promptWrite = await cli.readUntil(
        'Do you want to immediately run prettier',
      )
      expect(promptWrite).toMatch(
        'Do you want to immediately run prettier on all files in the project?',
      )
      await cli.type('y')

      const promptPrecommit = await cli.readUntil('precommit')
      expect(clearColorMarkers(promptPrecommit)).toMatchSnapshot()

      await cli.type('n')

      await cli.readUntil('Enjoy your configured workplace')

      expect(fs.readFileSync(writableFile).toString()).toMatchSnapshot()
    })
  })

  describe('running documents', () => {
    it('creates both files', async () => {
      const coc = rootPath('CODE_OF_CONDUCT.md')
      const license = rootPath('LICENSE')
      const returnLicense = reserveFile(license)
      const returnCOC = reserveFile(coc)

      returnFile = () => {
        returnLicense()
        returnCOC()
      }

      const cli = runOuxe(['documents', '--skip-install'])

      const promptDocument = await cli.readUntil(
        /Which documents do you want to create(.|\n)*/m,
      )
      expect(clearColorMarkers(promptDocument)).toMatchInlineSnapshot(`
        "? ü§î Which documents do you want to create? (Press <space> to select, <a> to tog
        gle all, <i> to invert selection)
        ‚ùØ‚óØ Code of Conduct
         ‚óØ License"
      `)

      await cli.type('a')

      const promptLicense = await cli.readUntil(/which license(.|\n)*/)
      expect(clearColorMarkers(promptLicense)).toMatchInlineSnapshot(`
        "? üìÑ  Please choose which license you want for your project: (Use arrow keys or 
        type to search)
        ‚ùØ 0BSD 
          AAL 
          AFL-1.1 
          AFL-1.2 
          AFL-2.0 
          AFL-2.1 
          AFL-3.0 
        (Move up and down to reveal more choices)"
      `)

      await cli.type('MIT')
      const selectedProject = await cli.readUntil('‚ùØ MIT')

      expect(clearColorMarkers(selectedProject)).toMatchInlineSnapshot(`
        "? üìÑ  Please choose which license you want for your project: MIT
        ‚ùØ MIT "
      `)

      await cli.type('')

      const promptUsername = await cli.readUntil(/name of the user(\n|.)*/)
      expect(clearColorMarkers(promptUsername)).toMatchInlineSnapshot(`
        "? üëì  What's the name of the user that'll sign the license (Yurick <yurick.hausc
        hild@gmail.com>) "
      `)

      // Press enter to confirm default
      await cli.type('')

      const promptEmail = await cli.readUntil(/provide an email.*/)
      expect(clearColorMarkers(promptEmail)).toMatchInlineSnapshot(
        `"? üìû Please provide an email for contact "`,
      )

      await cli.type('clifford@yurick.me')
      await cli.type('')

      await cli.readUntil(/Created file/)

      expect(fs.readFileSync(license).toString()).toMatchSnapshot()
      expect(fs.readFileSync(coc).toString()).toMatchSnapshot()
    })

    fit('creates a CODE_OF_CONDUCT.md', async () => {
      const codeOfConduct = rootPath('CODE_OF_CONDUCT.md')
      returnFile = reserveFile(codeOfConduct)

      const cli = runOuxe(['documents', '--skip-install'])

      const promptDocument = await cli.readUntil(/which documents(.|\n)*/i)
      expect(clearColorMarkers(promptDocument)).toMatchInlineSnapshot(`
        "? ü§î Which documents do you want to create? (Press <space> to select, <a> to tog
        gle all, <i> to invert selection)
        ‚ùØ‚óØ Code of Conduct
         ‚óØ License"
      `)

      // Press space to select Code of Conduct
      await cli.type(' ')

      const promptEmail = await cli.readUntil(/provide an email.*/)
      expect(clearColorMarkers(promptEmail)).toMatchInlineSnapshot(
        `"? üìû Please provide an email for contact "`,
      )

      await cli.type('clifford@yurick.me')
      await cli.readUntil(/Enjoy your configured workplace/)

      expect(fs.readFileSync(codeOfConduct).toString()).toMatchSnapshot()
    })

    it('creates a LICENSE file', async () => {
      const license = rootPath('LICENSE')

      returnFile = reserveFile(license)

      const cli = runOuxe(['documents', '--skip-install'])

      const promptDocument = await cli.readUntil(/which documents/i)
      expect(clearColorMarkers(promptDocument)).toMatchInlineSnapshot(`
        "? ü§î Which documents do you want to create? (Press <space> to select, <a> to tog
        gle all, <i> to invert selection)
        ‚ùØ‚óØ Code of Conduct
         ‚óØ License"
      `)

      // spacebar invert will select LICENSE (I'm yet to learn how to press down)
      await cli.type(' i')

      const promptLicense = await cli.readUntil(/which license/)
      expect(clearColorMarkers(promptLicense)).toMatchInlineSnapshot(`
        "? üìÑ  Please choose which license you want for your project: (Use arrow keys or 
        type to search)
        ‚ùØ 0BSD 
          AAL 
          AFL-1.1 
          AFL-1.2 
          AFL-2.0 
          AFL-2.1 
          AFL-3.0 
        (Move up and down to reveal more choices)"
      `)

      await cli.type('MIT')
      const selectedProject = await cli.readUntil(/(0BSD|Searching)/)

      expect(clearColorMarkers(selectedProject)).toMatchInlineSnapshot(`
        "? üìÑ  Please choose which license you want for your project: MIT
        ‚ùØ MIT "
      `)

      await cli.type('')

      const promptUsername = await cli.readUntil(/name of the user/)
      expect(clearColorMarkers(promptUsername)).toMatchInlineSnapshot(`
        "? üëì  What's the name of the user that'll sign the license (Yurick <yurick.hausc
        hild@gmail.com>) "
      `)

      // Press enter to confirm default
      await cli.type('')

      await cli.readUntil(/Enjoy your configured workplace/)

      expect(fs.readFileSync(license).toString()).toMatchSnapshot()
    })
  })
})
