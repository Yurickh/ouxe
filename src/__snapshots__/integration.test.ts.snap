// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ouxe running prettier writes all files 1`] = `
"export default function potato() {
  let a, b, c

  ;[a, b, c].forEach((x) => 0)
}
"
`;

exports[`ouxe runs help 1`] = `
"Usage: ouxe [configuration] [args]

Commands:
  ouxe            Choose from some configuration options               [default]
  ouxe prettier   Opinionated prettier configuration                [aliases: p]
  ouxe eslint     Opinionated eslint configuration                  [aliases: e]
  ouxe documents  Basic legal documents, like Code of Conduct and License
                                                                    [aliases: d]

Options:
  --version       Show version number                                  [boolean]
  --skip-install  Skip installation steps. Ideal for the unconnected.  [boolean]
  --help          Show help                                            [boolean]
"
`;
